### PR TITLE
[R-package] avoid unnecessary computation and add tests for Dataset set_reference() method

### PR DESCRIPTION
**Pull request by jameslamb**
Noticed while working on #4586.

`Dataset$set_reference()` allows changing the reference Dataset that a given `Dataset` is based on, which affects (for example) how features are binned.

Using `Dataset$set_reference()` to set `reference` to the existing object doesn't have any effect on the `Dataset` object. This PR proposes moving the existing check for that all the way up to the beginning of that method, to avoid any other unnecessary computation and slightly speed up this method for that case.

### CODE REVIEW

**Review by StrikerRUS — COMMENTED** [review]
Why do we leave `# Check for empty data` and `# Check for non-existing reference` untouched placed after setting `# Set known references`. I think it is misleading that some attributes are set (`categorical_feature`, `colnames` and `predictor`) but the whole method fails with fatal error. Doesn't it leave object in a corrupted, ok maybe inconsistent, state?

Moreover, I might be wrong but 
```
self$set_categorical_feature(categorical_feature = reference$.__enclos_env__$private$categorical_feature)
```
will simply crash without 
```
        # Reference is unknown
        if (!lgb.is.Dataset(reference)) {
          stop("set_reference: Can only use lgb.Dataset as a reference")
        }
```
placed before in case of wrong reference type.

**Comment by jameslamb** [comment]
Ah yes, you're right! Didn't notice that those should also be moved up sooner. I'll make that change here.

**Comment by jameslamb** [comment]
This refactoring is also making me realize that that method doesn't protect against the case where you run `set_reference(reference = NULL)`. Right now on `master`, that results in

```text
Error in self$set_colnames(colnames = reference$get_colnames()) : 
  attempt to apply non-function
```

I'll fix that here too

**Comment by jameslamb** [comment]
Ok, just moved these validations up further. I realized while doing this that there are currently no direct tests on `Dataset$set_reference()`, so added some of to be more confident in this change.

You can confirm that there are no direct tests on that method by running the following on `master`.

```shell
git grep -E "set.*ref" R-package/tests
```

File: R-package/tests/testthat/test_dataset.R [hunk 1]
```diff
@@ -74,6 +78,115 @@ test_that("Dataset$slice() supports passing Dataset attributes through '...'", {
   expect_identical(dsub1$getinfo("init_score"), init_score)
 })
 
+test_that("Dataset$set_reference() on a constructed datasetfails if raw data has been freed", {
```
Comment (StrikerRUS) [inline R-package/tests/testthat/test_dataset.R:81]:
```suggestion
test_that("Dataset$set_reference() on a constructed Dataset fails if raw data has been freed", {
```

File: R-package/tests/testthat/test_dataset.R [hunk 2]
```diff
@@ -74,6 +78,115 @@ test_that("Dataset$slice() supports passing Dataset attributes through '...'", {
...
+    , dtrain$.__enclos_env__$private$predictor$.__enclos_env__$private$handle
+  )
+  expect_identical(
+    dtest$.__enclos_env__$private$categorical_feature
+    , dtrain$.__enclos_env__$private$categorical_feature
+  )
+  expect_identical(
+    dtest$get_colnames()
+    , dtrain$get_colnames()
+  )
```
Comment (StrikerRUS) [inline R-package/tests/testthat/test_dataset.R:187]: Should we also check that column names differ from `test_original_feature_names` at this point?

**Review by StrikerRUS — APPROVED** [review]
With this PR, the order of code lines in `set_reference()` function looks good to me, thanks!
Just two minor comments regarding new tests.

File: R-package/tests/testthat/test_dataset.R (hunk 2 above)
Comment (jameslamb) [reply R-package/tests/testthat/test_dataset.R:187]: good idea! That would make this a bit safer. Added:

```suggestion
  expect_identical(
    dtest$get_colnames()
    , dtrain$get_colnames()
  )
  expect_false(
    identical(dtest$get_colnames(), test_original_feature_names)
  )
```

**Pull request merged**